src/System.Buffers.ReaderWriter/System/Buffers/Reader/BufferReader_text.cs

```diff
@@ -58,42 +64,88 @@ private unsafe static bool TryParseSlow(ref BufferReader<byte> reader, out bool
         /// Try to parse an int out of the reader (as UTF-8).
         /// </summary>
         /// <param name="value">The parsed int value or default if failed.</param>
+        /// <param name="standardFormat">Expected format.</param>
+        /// <remarks>
+        /// <see cref="Utf8Parser.TryParse(ReadOnlySpan{byte}, out int, out int, char)"/> for supported formats.
+        /// </remarks>
+        /// <exception cref="FormatException">Thrown if the given format isn't supported.</exception>
         /// <returns>True if successfully parsed.</returns>
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
-        public static unsafe bool TryParse(ref this BufferReader<byte> reader, out int value)
+        public static unsafe bool TryParse(ref this BufferReader<byte> reader, out int value, char standardFormat = '\0')
         {
             ReadOnlySpan<byte> unread = reader.UnreadSpan;
-            if (Utf8Parser.TryParse(unread, out value, out int consumed) && consumed < unread.Length)
+            if (!Utf8Parser.TryParse(unread, out value, out int consumed, standardFormat))
+            {
+                return false;
+            }
+
+            if (consumed < unread.Length)
             {
+                // The parser found a digit it wouldn't consume so we have all valid data.
                 reader.Advance(consumed);
                 return true;
             }
 
-            return TryParseSlow(ref reader, out value);
+            // If we ate all of our unread there may be more valid digits
+            return TryParseSlow(ref reader, out value, standardFormat);
         }
 
-        private static unsafe bool TryParseSlow(ref BufferReader<byte> reader, out int value)
+        private static unsafe bool TryParseSlow(ref BufferReader<byte> reader, out int value, char standardFormat)
         {
-            const int MaxLength = 15;
-
-            ReadOnlySpan<byte> unread = reader.UnreadSpan;
+            // Note that ints can have leading zeros ("-00000000000000127") which makes predicting the max
+            // length difficult. Typically the int will be less than 15 characters ("-2,147,483,648") so we'll
+            // work in those increments.
+            int bufferSize = 15;
 
-            if (unread.Length > MaxLength)
+            // Try using the stack for scratch space.
+            do
```

Comment:
Rather than continuously doubling (and potentially throwing if there are 2.1 billion leading zeroes, which should be valid), why not skip all leading zeroes ahead of time? In the common case, it will be a single check but the implementation would be clearer.

Reply:
We'd have to deal with '+', '-', '.' and other formats that might come in the future.  Doesn't seem wise to couple that closely and add that much complexity for something that arguably should be handled by the consumer of this API.

That said, I think might even be more aggressive here. Lets say someone throws an empty 1GB buffer at us- should we really cruise the entire thing and return 0? Is that really what a user would want?